File: 17-week-13/01-day-3-to-5--add-two-numbers-list-nodes/solutions/addTwoList.rb
class ListNode
    attr_accessor :val, :next
    def initialize(val = 0, _next = nil)
        @val = val
        @next = _next
    end
end

def add_two_numbers(l1, l2)
    n1, n2 = l1, l2
    
    current = answer = ListNode.new(nil)
    carry = 0
    
    while n1 || n2 || 0 < carry
        a = n1 ? n1.val : 0
        b = n2 ? n2.val : 0
        x = a + b + carry
        
        carry = x / 10
        x -= 10 if 0 < carry
        
        node = ListNode.new(x)
        current&.next = node
        current = node
        
        n1, n2 = n1&.next, n2&.next
    end 
    answer.next
end
  

if __FILE__ == $PROGRAM_NAME
puts "Expecting: 0"
l1 = ListNode.new(0)
l2 = ListNode.new(0)
puts "=>", add_two_numbers(l1,l2)

# Don't forget to add your own!
end
  
# More details here: https://leetcode.com/problems/add-two-numbers/solution/

File: 17-week-13/01-day-3-to-5--add-two-numbers-list-nodes/solutions/addTwoList.js
class Node {
    constructor(val = null, next = null) {
      this.val = val;
      this.next = next;
    }
}

function addTwoList(l1, l2) {

let node = null

const carry = arguments[2]
if (l1 || l2) {
    const val1 = l1 ? l1.val : 0
    const val2 = l2 ? l2.val : 0
    const next1 = l1 ? l1.next : null
    const next2 = l2 ? l2.next : null
    const val = carry ? val1 + val2 + 1 : val1 + val2
    node = new Node(val % 10)
    node.next = addTwoList(next1, next2, val >= 10)  
} else if (carry) {
    node = new Node(1)
    node.next = null
}
console.log(node)
return node

}

if (require.main === module) {
// add your own tests in here
console.log("Expecting: { val: 0, next: null }");
console.log("=>", addTwoNumbers({ val: 0, next: null }, { val: 0, next: null }));

}

module.exports = {
    Node,
    addTwoList
};

// More details here: https://leetcode.com/problems/add-two-numbers/solution/